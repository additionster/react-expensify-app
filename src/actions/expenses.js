import {v7 as uuid} from 'uuid';
import { db, schema } from '../firebase/firebase';
import { ref, push, get, remove, update } from 'firebase/database';


//ADD expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        return push(ref(db, `users/${uid}/${schema}`), expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }));
            });
    };
};

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return update(ref(db, `users/${uid}/${schema}/${id}`), {
            ...updates
        }).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
});

export const startRemoveExpense = ({ id }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(db, `users/${uid}/${schema}/${id}`))
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    };
};

//SET_EXPENSE
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const expenses = [];
        return get(ref(db, `users/${uid}/${schema}`))
            .then((snapshot) => {
                snapshot.forEach(childSnapshot => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                dispatch(setExpenses(expenses));
            });
    };
};