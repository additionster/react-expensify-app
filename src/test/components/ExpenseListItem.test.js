import React from "react";
import ExpenseListItem from "../../components/ExpenseListItem";
import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import { expenses } from "../fixtures/expenses";
import { BrowserRouter } from "react-router";

test('testRenderExpenseListItem', () => {
    const renderer = render(<BrowserRouter><ExpenseListItem {...expenses[0]} /></BrowserRouter>);
    expect(renderer.baseElement).toMatchSnapshot();
});