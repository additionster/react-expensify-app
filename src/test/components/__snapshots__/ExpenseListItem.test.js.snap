// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`testRenderExpenseListItem 1`] = `
<body>
  <div>
    <div>
      <h3>
        <a
          data-discover="true"
          href="/edit/1"
        >
          Gum
        </a>
      </h3>
      <p>
        195
         - 
        0
      </p>
    </div>
  </div>
</body>
`;
